File: LibraryBuildProject/helucollapsingtabbar/src/main/java/cz/helu/helucollapsingtabbar/HeluCollapsingTabBar.java
package cz.helu.helucollapsingtabbar;

import android.animation.LayoutTransition;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;

import java.util.ArrayList;
import java.util.List;


public class HeluCollapsingTabBar extends LinearLayout
{
	private List<HeluCollapsingTabBarButton> mButtonsList;
	private int mButtonSize = 0;
	private int mButtonPadding = 0;
	private int mButtonSpacing = 0;
	private Drawable mBackground;
	private boolean mIsCollapsed = false;
	private int mSelectedItem = 0;


	public HeluCollapsingTabBar(Context context)
	{
		this(context, null);
	}


	public HeluCollapsingTabBar(Context context, AttributeSet attrs)
	{
		super(context, attrs);
	}


	public HeluCollapsingTabBar(Builder builder)
	{
		super(builder.context);
		initFromBuilder(builder);
	}


	public void initFromBuilder(Builder builder)
	{
		this.mBackground = builder.background;
		this.mButtonsList = builder.buttonsList;
		this.mButtonSize = builder.buttonSize;
		this.mButtonPadding = builder.buttonPadding;
		this.mButtonSpacing = builder.buttonSpacing;

		initView();
	}


	public void setSelectedItem(int position)
	{
		if(position >= getChildCount())
			return;

		mSelectedItem = position;

		collapseView();
	}


	public void collapseView()
	{
		if(!(getChildAt(mSelectedItem) instanceof ImageView))
			return;

		ImageView view = (ImageView) getChildAt(mSelectedItem);

		for(int i = 0; i < getChildCount(); i++)
		{
			if(!(getChildAt(i) instanceof ImageView))
				continue;

			ImageView child = (ImageView) getChildAt(i);
			if(!child.equals(view))
			{
				child.setVisibility(View.GONE);
			}
			else
			{
				child.setImageDrawable(mButtonsList.get(i).getIcon());
				LayoutParams params = (LayoutParams) view.getLayoutParams();
				params.gravity = Gravity.CENTER;
				clearViewMargins(params, view);
				mSelectedItem = i;
			}
		}
		mIsCollapsed = true;
	}


	public void expandView()
	{
		if(!(getChildAt(mSelectedItem) instanceof ImageView))
			return;

		ImageView view = (ImageView) getChildAt(mSelectedItem);
		LayoutParams params = (LayoutParams) view.getLayoutParams();
		params.gravity = Gravity.CENTER;
		setupViewMargins(params, view);

		for(int i = 0; i < getChildCount(); i++)
		{
			if(!(getChildAt(i) instanceof ImageView))
				continue;

			ImageView child = (ImageView) getChildAt(i);
			child.setVisibility(View.VISIBLE);

			if(i != mSelectedItem)
				child.setImageDrawable(mButtonsList.get(i).getInActiveIcon());
		}

		mIsCollapsed = false;
	}


	@SuppressWarnings("unused")
	public boolean isCollapsed()
	{
		return mIsCollapsed;
	}


	private void initView()
	{
		// Set Parent layout background drawable
		setBackground(mBackground);
		setLayoutTransition(new LayoutTransition());

		// Setup all buttons view and add them into layout
		for(final HeluCollapsingTabBarButton button : mButtonsList)
		{
			final ImageView buttonImageView = new ImageView(getContext());

			LayoutParams params = new LayoutParams(mButtonSize, mButtonSize);
			params.gravity = Gravity.CENTER;
			setupViewMargins(params, buttonImageView);

			buttonImageView.setLayoutParams(params);
			buttonImageView.setPadding(mButtonPadding, mButtonPadding, mButtonPadding, mButtonPadding);
			buttonImageView.setImageDrawable(button.getIcon());
			buttonImageView.setOnClickListener(new OnClickListener()
			{
				@Override
				public void onClick(View view)
				{
					mSelectedItem = indexOfChild(view);
					onButtonClicked();
				}
			});

			addView(buttonImageView);
		}
	}


	private void onButtonClicked()
	{
		if(mIsCollapsed)
		{
			expandView();
		}
		else
		{
			ImageView view = (ImageView) getChildAt(mSelectedItem);
			mButtonsList.get(mSelectedItem).getOnClickListener().onClick(view);
			collapseView();
		}
	}


	private void setupViewMargins(LayoutParams params, ImageView view)
	{
		// Set margin for view, depending if the layout is vertical or horizontal
		if(getOrientation() == HORIZONTAL)
			params.setMargins(mButtonSpacing, 0, mButtonSpacing, 0);
		else
			params.setMargins(0, mButtonSpacing, 0, mButtonSpacing);

		view.setLayoutParams(params);
	}


	private void clearViewMargins(LayoutParams params, ImageView view)
	{
		params.setMargins(0, 0, 0, 0);
		view.setLayoutParams(params);
	}


	public static class Builder
	{
		private final int DEFAULT_BUTTON_SIZE = 56;
		private final int DEFAULT_BUTTON_PADDING = 16;
		private final int DEFAULT_BUTTON_SPACING = 12;

		private Context context;
		private List<HeluCollapsingTabBarButton> buttonsList = new ArrayList<>();
		private int buttonSize = 0;
		private int buttonPadding = 0;
		private int buttonSpacing = 0;
		private Drawable background;


		public Builder(Context context)
		{
			this.context = context;

			buttonSize = convertDpToPx(DEFAULT_BUTTON_SIZE);
			buttonPadding = convertDpToPx(DEFAULT_BUTTON_PADDING);
			buttonSpacing = convertDpToPx(DEFAULT_BUTTON_SPACING);
		}


		@SuppressWarnings("unused")
		public Builder withBackground(Drawable background)
		{
			this.background = background;
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withButtonSize(int dimension)
		{
			this.buttonSize = context.getResources().getDimensionPixelSize(dimension);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withButtonPadding(int dimension)
		{
			this.buttonPadding = context.getResources().getDimensionPixelSize(dimension);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withButtonSpacing(int dimension)
		{
			this.buttonSpacing = context.getResources().getDimensionPixelSize(dimension);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withButtonSizePx(int dimension)
		{
			this.buttonSize = dimension;
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withButtonPaddingPx(int dimension)
		{
			this.buttonPadding = dimension;
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withButtonSpacingPx(int dimension)
		{
			this.buttonSpacing = dimension;
			return this;
		}


		@SuppressWarnings("unused")
		public Builder addButton(HeluCollapsingTabBarButton button)
		{
			this.buttonsList.add(button);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder addButton(Drawable activeIcon, OnClickListener clickListener)
		{
			HeluCollapsingTabBarButton button = new HeluCollapsingTabBarButton();
			button.setIcon(activeIcon);
			button.setInActiveIcon(activeIcon);
			button.setOnClickListener(clickListener);
			this.buttonsList.add(button);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder addButton(Drawable activeIcon, Drawable inactiveIcon, OnClickListener clickListener)
		{
			HeluCollapsingTabBarButton button = new HeluCollapsingTabBarButton();
			button.setIcon(activeIcon);
			button.setInActiveIcon(inactiveIcon);
			button.setOnClickListener(clickListener);
			this.buttonsList.add(button);
			return this;
		}


		@SuppressWarnings("unused")
		public HeluCollapsingTabBar build()
		{
			return new HeluCollapsingTabBar(this);
		}


		int convertDpToPx(int dp)
		{
			return Math.round(dp * context.getResources().getDisplayMetrics().xdpi / DisplayMetrics.DENSITY_DEFAULT);
		}
	}
}

File: LibraryBuildProject/helucollapsingtabbar/src/main/java/cz/helu/helucollapsingtabbar/HeluCollapsingTabBar.kt
package cz.helu.helucollapsingtabbar

import android.animation.LayoutTransition
import android.content.Context
import android.graphics.drawable.Drawable
import android.support.v4.content.ContextCompat
import android.util.AttributeSet
import android.util.DisplayMetrics
import android.view.Gravity
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import cz.helu.helucollapsingtabbar.entity.HeluCollapsingTabBarButton
import java.util.*

@Suppress("MemberVisibilityCanBePrivate", "unused")
class HeluCollapsingTabBar : LinearLayout {
    private var buttonsList: List<HeluCollapsingTabBarButton> = ArrayList()
    private var buttonSize = 0
    private var buttonPadding = 0
    private var buttonSpacing = 0
    private var bgDrawable: Drawable? = null
    var isCollapsed = false
        private set
    private var selectedPosition = 0

    @JvmOverloads
    constructor(context: Context, attrs: AttributeSet? = null) : super(context, attrs)

    constructor(builder: Builder) : super(builder.context) {
        initFromBuilder(builder)
    }

    fun initFromBuilder(builder: Builder) {
        this.bgDrawable = builder.background
        this.buttonsList = builder.buttonsList
        this.buttonSize = builder.buttonSize
        this.buttonPadding = builder.buttonPadding
        this.buttonSpacing = builder.buttonSpacing

        initView()
    }

    fun setSelectedPosition(position: Int) {
        if (position >= childCount)
            return

        selectedPosition = position

        collapseView()
    }

    fun getSelectedPosition(): Int = selectedPosition

    fun collapseView() {
        if (getChildAt(selectedPosition) !is ImageView)
            return

        val view = getChildAt(selectedPosition) as ImageView

        for (i in 0 until childCount) {
            if (getChildAt(i) !is ImageView)
                continue

            val child = getChildAt(i) as ImageView
            if (child != view) {
                child.visibility = View.GONE
            } else {
                child.setImageDrawable(buttonsList[i].icon)
                val params = view.layoutParams as LinearLayout.LayoutParams
                params.gravity = Gravity.CENTER
                clearViewMargins(params, view)
                selectedPosition = i
            }
        }
        isCollapsed = true
    }

    fun expandView() {
        if (getChildAt(selectedPosition) !is ImageView)
            return

        val view = getChildAt(selectedPosition) as ImageView
        val params = view.layoutParams as LinearLayout.LayoutParams
        params.gravity = Gravity.CENTER
        setupViewMargins(params, view)

        for (i in 0 until childCount) {
            if (getChildAt(i) !is ImageView)
                continue

            val child = getChildAt(i) as ImageView
            child.visibility = View.VISIBLE

            if (i != selectedPosition)
                child.setImageDrawable(buttonsList[i].inactiveIcon)
        }

        isCollapsed = false
    }

    private fun initView() {
        // Set Parent layout background drawable
        background = bgDrawable
        layoutTransition = LayoutTransition()

        // Setup all buttons view and add them into layout
        for (button in buttonsList) {
            val buttonImageView = ImageView(context)

            val params = LinearLayout.LayoutParams(buttonSize, buttonSize)
            params.gravity = Gravity.CENTER
            setupViewMargins(params, buttonImageView)

            buttonImageView.layoutParams = params
            buttonImageView.setPadding(buttonPadding, buttonPadding, buttonPadding, buttonPadding)
            buttonImageView.setImageDrawable(button.icon)
            buttonImageView.setOnClickListener { view ->
                selectedPosition = indexOfChild(view)
                onButtonClicked()
            }

            addView(buttonImageView)
        }
    }

    private fun onButtonClicked() {
        if (isCollapsed) {
            expandView()
        } else {
            val view = getChildAt(selectedPosition) as ImageView
            buttonsList[selectedPosition].onClickListener?.onClick(view)
            collapseView()
        }
    }

    private fun setupViewMargins(params: LinearLayout.LayoutParams, view: ImageView) {
        // Set margin for view, depending if the layout is vertical or horizontal
        if (orientation == LinearLayout.HORIZONTAL)
            params.setMargins(buttonSpacing, 0, buttonSpacing, 0)
        else
            params.setMargins(0, buttonSpacing, 0, buttonSpacing)

        view.layoutParams = params
    }

    private fun clearViewMargins(params: LinearLayout.LayoutParams, view: ImageView) {
        params.setMargins(0, 0, 0, 0)
        view.layoutParams = params
    }

    class Builder(val context: Context) {
        companion object {
            private const val DEFAULT_BUTTON_SIZE = 48
            private const val DEFAULT_BUTTON_PADDING = 12
            private const val DEFAULT_BUTTON_SPACING = 16
        }

        val buttonsList = ArrayList<HeluCollapsingTabBarButton>()
        var buttonSize = 0
        var buttonPadding = 0
        var buttonSpacing = 0
        var background: Drawable? = null

        init {
            buttonSize = convertDpToPx(DEFAULT_BUTTON_SIZE)
            buttonPadding = convertDpToPx(DEFAULT_BUTTON_PADDING)
            buttonSpacing = convertDpToPx(DEFAULT_BUTTON_SPACING)
            background = ContextCompat.getDrawable(context, R.drawable.shape_collapsing_tab_bar)
        }

        fun withBackground(background: Drawable): Builder {
            this.background = background
            return this
        }

        fun withButtonSize(dimension: Int): Builder {
            this.buttonSize = context.resources.getDimensionPixelSize(dimension)
            return this
        }

        fun withButtonPadding(dimension: Int): Builder {
            this.buttonPadding = context.resources.getDimensionPixelSize(dimension)
            return this
        }

        fun withButtonSpacing(dimension: Int): Builder {
            this.buttonSpacing = context.resources.getDimensionPixelSize(dimension)
            return this
        }

        fun withButtonSizePx(dimension: Int): Builder {
            this.buttonSize = dimension
            return this
        }

        fun withButtonPaddingPx(dimension: Int): Builder {
            this.buttonPadding = dimension
            return this
        }

        fun withButtonSpacingPx(dimension: Int): Builder {
            this.buttonSpacing = dimension
            return this
        }

        fun addButton(button: HeluCollapsingTabBarButton): Builder {
            this.buttonsList.add(button)
            return this
        }

        fun addButton(activeIcon: Drawable, clickListener: View.OnClickListener): Builder {
            val button = HeluCollapsingTabBarButton()
            button.icon = activeIcon
            button.inactiveIcon = activeIcon
            button.onClickListener = clickListener
            this.buttonsList.add(button)
            return this
        }

        fun addButton(activeIcon: Drawable, inactiveIcon: Drawable, clickListener: View.OnClickListener): Builder {
            val button = HeluCollapsingTabBarButton()
            button.icon = activeIcon
            button.inactiveIcon = inactiveIcon
            button.onClickListener = clickListener
            this.buttonsList.add(button)
            return this
        }

        fun build(): HeluCollapsingTabBar {
            return HeluCollapsingTabBar(this)
        }

        internal fun convertDpToPx(dp: Int): Int {
            return Math.round(dp * context.resources.displayMetrics.xdpi / DisplayMetrics.DENSITY_DEFAULT)
        }
    }
}